### DISCUSSION TITLE
Error closing the app in Windows

### DISCUSSION THREAD

**Discussion by splcoder** [beeware/toga#2169] [Debugging help]
Hi there

I'm new in Beeware, and I'm doing some examples from the Beeware home page, more exactly "https://docs.beeware.org/en/latest/tutorial/tutorial-2.html", and I'm having the next error when the **app opens, click the button, and then close the app** (but NOT when the app opens, and directly then close the app):
```
_Windows fatal exception: access violation

Current thread 0x00004a10 (most recent call first):
  <no Python frame>

Thread 0x000028c4 (most recent call first):
  File "C:\Installed\Docs\Anaconda\envs\beeware\lib\site-packages\clr_loader\types.py", line 64 in __call__
  File "C:\Installed\Docs\Anaconda\envs\beeware\lib\site-packages\pythonnet\__init__.py", line 161 in unload_
```
Greetings

**Comment by freakboy3742** [comment]
I can't say I've seen that error myself - can you clarify exactly what you're doing to get this error?

1. Click on "the button" - Tutorial 2 has 10 buttons on the right hand side panel, plus 4 toolbar buttons. Which button do you mean?
2. "Directly close the app" - using what mechanism? The close widget on the top right of the window? File > Exit from the menu? Something else
3. Some of the button handlers have long running effects in the terminal. Are you quitting the app immediately after pressing the button, or waiting some time for the handler to finish running (i.e., what else is displayed in the terminal other than the stack trace)

Also - can you tell us:
* The version of Toga you're using
* The version of Python you're using
* The version of Windows you're using

**Reply by rmartin16** [reply]
as a datapoint....I could only repro this using Toga 0.3.1. In those cases, I used `briefcase dev`, clicked the button, and clicked the `X` in the top right. Although, I had a stack line under "Current thread" for `winforms.proacter.tick()`. Seemed to be a low chance of happening, though....had to repeat this process many times.

As a note, I didn't use Tutorial 2....I used the "print Hello, \<name\>" example in the BeeWare tutorial.

**Reply by rmartin16** [reply]
Tried ~100 times with Toga `main` branch installed but still couldn't repro.

Was able to pretty easily with Toga 0.3.1, though....this time without a mention to `winforms.proacter.tick()`...so maybe a red herring.

<details>
<summary>Log</summary>

```
Date/Time:       2023-10-23 04:58:42 
Command line:    C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase\Scripts\briefcase dev --log

OS Release:      Windows 10
OS Version:      10.0.22621
Architecture:    AMD64
Platform:        Windows-10-10.0.22621-SP0

Python exe:      C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase\Scripts\python.exe
Python version:  3.10.11 (tags/v3.10.11:7d4cc5a, Apr  5 2023, 00:38:17) [MSC v.1929 64 bit (AMD64)]
Virtual env:     True
Conda env:       False

Briefcase:       0.3.16
Target platform: windows
Target format:   None

Environment Variables:
    ALLUSERSPROFILE=C:\ProgramData
    APPDATA=C:\Users\user\AppData\Roaming
    COMMONPROGRAMFILES=C:\Program Files\Common Files
    COMMONPROGRAMFILES(X86)=C:\Program Files (x86)\Common Files
    COMMONPROGRAMW6432=C:\Program Files\Common Files
    COMPUTERNAME=VM-WIN11
    COMSPEC=C:\Windows\system32\cmd.exe
    DRIVERDATA=C:\Windows\System32\Drivers\DriverData
    HOMEDRIVE=C:
    HOMEPATH=\Users\user
    LOCALAPPDATA=C:\Users\user\AppData\Local
    LOGONSERVER=\\VM-WIN11
    NUMBER_OF_PROCESSORS=4
    ONEDRIVE=C:\Users\user\OneDrive
    OS=Windows_NT
    PATH=C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase\Scripts;C:\Program Files\PowerShell\7;C:\Users\user\.pyenv\pyenv-win\shims;C:\Users\user\.pyenv\pyenv-win\versions\3.9.13;C:\Users\user\.pyenv\pyenv-win\versions\3.8.10;C:\Users\user\.pyenv\pyenv-win\versions\3.10.11;C:\Users\user\.pyenv\pyenv-win\versions\3.11.3;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Git\cmd;;C:\Program Files\dotnet\;C:\Program Files\starship\bin\;C:\Program Files\Docker\Docker\resources\bin;C:\Program Files\PowerShell\7\;C:\Users\user\.pyenv\pyenv-win\bin;C:\Users\user\.pyenv\pyenv-win\shims;C:\Users\user\AppData\Local\Microsoft\WindowsApps;C:\Users\user\.dotnet\tools;C:\Users\user\.dotnet\tools;C:\Users\user\AppData\Local\Programs\Microsoft VS Code\bin
    PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
    POWERSHELL_DISTRIBUTION_CHANNEL=MSI:Windows 10 Pro
    PROCESSOR_ARCHITECTURE=AMD64
    PROCESSOR_IDENTIFIER=AMD64 Family 25 Model 33 Stepping 0, AuthenticAMD
    PROCESSOR_LEVEL=25
    PROCESSOR_REVISION=2100
    PROGRAMDATA=C:\ProgramData
    PROGRAMFILES=C:\Program Files
    PROGRAMFILES(X86)=C:\Program Files (x86)
    PROGRAMW6432=C:\Program Files
    PSMODULEPATH=C:\Users\user\Documents\PowerShell\Modules;C:\Program Files\PowerShell\Modules;c:\program files\powershell\7\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\Windows\system32\WindowsPowerShell\v1.0\Modules
    PUBLIC=C:\Users\Public
    PYENV=C:\Users\user\.pyenv\pyenv-win\
    PYENV_HOME=C:\Users\user\.pyenv\pyenv-win\
    PYENV_ROOT=C:\Users\user\.pyenv\pyenv-win\
    SESSIONNAME=Console
    STARSHIP_SESSION_KEY=********************
    STARSHIP_SHELL=pwsh
    SYSTEMDRIVE=C:
    SYSTEMROOT=C:\Windows
    TEMP=C:\Users\user\AppData\Local\Temp
    TMP=C:\Users\user\AppData\Local\Temp
    USERDOMAIN=vm-win11
    USERDOMAIN_ROAMINGPROFILE=vm-win11
    USERNAME=user
    USERPROFILE=C:\Users\user
    VIRTUAL_ENV=C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase
    VIRTUAL_ENV_DISABLE_PROMPT=1
    WINDIR=C:\Windows
    WSLENV=WT_SESSION:WT_PROFILE_ID:
    WT_PROFILE_ID={574e775e-4f2a-5b96-ac1e-a2962a402336}
    WT_SESSION=8d6af7a9-c23d-4ccf-9664-861d7fd60cb6
    _OLD_VIRTUAL_PATH=C:\Program Files\PowerShell\7;C:\Users\user\.pyenv\pyenv-win\shims;C:\Users\user\.pyenv\pyenv-win\versions\3.9.13;C:\Users\user\.pyenv\pyenv-win\versions\3.8.10;C:\Users\user\.pyenv\pyenv-win\versions\3.10.11;C:\Users\user\.pyenv\pyenv-win\versions\3.11.3;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files\Git\cmd;;C:\Program Files\dotnet\;C:\Program Files\starship\bin\;C:\Program Files\Docker\Docker\resources\bin;C:\Program Files\PowerShell\7\;C:\Users\user\.pyenv\pyenv-win\bin;C:\Users\user\.pyenv\pyenv-win\shims;C:\Users\user\AppData\Local\Microsoft\WindowsApps;C:\Users\user\.dotnet\tools;C:\Users\user\.dotnet\tools;C:\Users\user\AppData\Local\Programs\Microsoft VS Code\bin

Briefcase Log:
[04:58:38]                                                                                                                                                                dev.py:228
           [helloworld] Starting in dev mode...                                                                                                                           dev.py:228
                                                                                                                                                                   subprocess.py:711
           >>> Running Command:                                                                                                                                    subprocess.py:712
           >>>     'C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase\Scripts\python.exe' -c 'import runpy,                                               subprocess.py:713
           sys;sys.path.pop(0);sys.argv.extend([]);runpy.run_module("helloworld", run_name="__main__", alter_sys=True)'                                                             
           >>> Working Directory:                                                                                                                                  subprocess.py:720
           >>>     C:\Users\user                                                                                                                                   subprocess.py:721
           >>> Environment Overrides:                                                                                                                              subprocess.py:730
           >>>     PYTHONPATH=C:\Users\user\github\beeware\briefcase\tmp\helloworld\src                                                                            subprocess.py:732
           >>>     PYTHONMALLOC=default                                                                                                                            subprocess.py:732
           >>>     PYTHONUNBUFFERED=1                                                                                                                              subprocess.py:732
           >>>     PYTHONDEVMODE=1                                                                                                                                 subprocess.py:732
           >>>     PYTHONUTF8=1                                                                                                                                    subprocess.py:732
           ===========================================================================                                                                                    run.py:160
[04:58:41] Hello,                                                                                                                                                  subprocess.py:681
[04:58:42] Windows fatal exception: access violation                                                                                                               subprocess.py:681
                                                                                                                                                                   subprocess.py:681
           Current thread 0x00000d44 (most recent call first):                                                                                                     subprocess.py:681
             <no Python frame>                                                                                                                                     subprocess.py:681
                                                                                                                                                                   subprocess.py:681
           Thread 0x00001a74 (most recent call first):                                                                                                             subprocess.py:681
             File "C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase\lib\site-packages\clr_loader\types.py", line 64 in __call__                          subprocess.py:681
             File "C:\Users\user\github\beeware\briefcase\venv-3.10-briefcase\lib\site-packages\pythonnet\__init__.py", line 157 in unload                         subprocess.py:681
```

</details>

**Comment by splcoder** [comment]
I realised that this error is really extrange:
	Yesterday it gaves me many errors of this type.
	Today only 1 time at first hour.

I'm working with:
	Windows	10
	Python	3.10.13
	Toga	0.3.1

Executed with:
	briefcase dev

Adding a note: it is not required to do anything, only close the app within "X" (I have had more errors of this type).

**Comment by aendie** [comment]
Greetings!
I found this discussion because I apparently have the same error:

<img width="1092" height="376" alt="Screenshot 2025-11-06 211327" src="https://github.com/user-attachments/assets/a2c8a843-428b-4577-9965-2ee6bcd33ffa" />

CPU:    AMD Ryzen 7 9700X 8-Core Processor (3.80 GHz); 
RAM:   32,0 GB (31,1 GB usable)
OS:      Windows 11 Pro 25H2, OS Build 26200.7019
Type:   64-bit operating system, x64-based processor
Python: 3.13.0
Toga:   0.5.2

The error appears to be random: whether it is closed with File-Exit... or clicking on the "X" (top-right of window).
It usually fails though.
It **DOES NOT FAIL** if the app is closed with Ctrl-C in the Command Prompt.

App code:

```
import toga
from toga.style.pack import COLUMN, ROW
import random

def get_coordinates():
    coordinate  = random.randint(1000,9999)
    coordinate2 = random.randint(1000,9999)
    total = str(coordinate) + ' / ' + str(coordinate2)
    return(total)

class TargetCoordinates(toga.App):

    def startup(self):
        main_box = toga.Box(direction=COLUMN)

        name_label = toga.Label(
            "Target coordinates:  " + get_coordinates(),
            margin=(0, 5),
        )

        name_box = toga.Box(direction=ROW, margin=5)
        name_box.add(name_label)
        main_box.add(name_box)

        self.main_window = toga.MainWindow(title='Random Target Coordinates Generator.')
        self.main_window.content = main_box
        self.main_window.show()

def main():
    return TargetCoordinates()
```

**Reply by freakboy3742** [reply]
Thanks for those additional details. I haven't been able to reproduce the problem myself; but my test machine is still on Windows 10. - I don't know if might be related to that.

The other possibility is that it's an issue with asyncio support in 3.13.0 - Python .0 releases are the ones most likely to have issues, so upgrading to a more recent 3.13 release (or any other Python release that isn't a .0) would be worth a try. I'm not aware of anything *specifically* in 3.13.0 that would cause the problem you describe, though.

If that doesn't work - two additional diagnostic details that might be helpful:
1. Do you see the same problem under `briefcase run`? Or is it just `briefcase dev`?
2. Do you have any additional .NET tooling installed? If so, what versions?

**Comment by aendie** [comment]
Thanks for your response.

- The problem also occurs with the standard BeeWare tutorial - it has nothing specifically to do with my app.
- It only occurs with `briefcase dev` - not with `briefcase run`

I upgraded Python 3.13 to 3.13.9. I don't use virtualenv or pyenv: I only have one system Python version installed.

- It fails frequently when closing using the 'X' button (top-right of window)
- It fails rarely when closing with File-Exit (maybe on 7th try)

The following are installed on my Windows system:

- Microsoft Visual C++ 2008 Redistributable - x86 9.0.30729.6161
- Microsoft Visual C++ 2015-2022 Redistributable (x64) - 14.36.32532
- Microsoft Visual C++ 2015-2022 Redistributable (x64) - 14.38.33135
- Microsoft Windows Desktop Runtime - 8.0.20(x64)

Not sure if I need these still. (I believe the last one above was required by the Handbrake video converter 1.10.2 on Windows 11.)

I wonder if simplifying my environment PATH variable might have any effect?

**Comment by aendie** [comment]
I got hold of an ASUS Zenbook runing on an Intel Core i7-8565U CPU with Windows 11. I installed Python 3.13.9 and ran the BeeWare tutorial. It ran without error (when closing it in the app)!

I next suspected Microsoft Windows Desktop Runtime - 8.0.20(x64) (= Microsoft .NET 8.0 Runtime).
**Uninstalling** it on my desktop showed that the HandBrake 1.10.2 video converter no longer worked, however I still had the `Windows fatal exception: access violation` on closing helloworld. Also, I installed the Microsoft .NET 8.0 Runtime 8.0.20 on the Zenbook - it still runs without any error. **Installing** the latest Microsoft .NET 8.0 Runtime 8.0.21 on the desktop also had no effect (except that HandBrake again runs).

Next I uninstalled the two Visual C++ 14.* Redistributables ..... and the desktop still failed. (The Zenbook also has the Visual C++ 2008 Redistributable 9.0.30729.6161, so I left it installed.)

Next difference: the Zenbook was on Windows 11 24H2 whereas my desktop is on 25H2. So I upgraded Windows 11 on the Zenbook. Again, no difference in the symptoms.

I hope to discover something that makes my desktop work - or the Zenbook fail (when closing the BeeWare tutorial).

**Reply by freakboy3742** [reply]
Thanks for the update. 

The fact that this is only happening with `briefcase dev` leads me to suspect it might be an issue with Python "dev mode". `briefcase dev` runs app with the `-X dev` flag when in dev mode, because that can be good at exposing edge case bugs - but in this case, based on the location reported in the error log, it isn't actually your problem to fix. It looks like an issue with Python.net shutting down. 

The next release of Briefcase (due out very shortly) will have the ability to disable dev mode. However, if you're up for some experimentation before then: it would be interesting to know if disabling line 35 of `briefcase/commands/dev.py` (manually editing the file in your virtual environment) fixes the problem.

**Comment by aendie** [comment]
Greetings! 
Well, I commented out line 35 in `briefcase/commands/dev.py`. Just to verify - this is **before** I commented it out:

<img width="1032" height="422" alt="Screenshot 2025-11-11 164709" src="https://github.com/user-attachments/assets/22141914-993a-4e28-8fc7-b927f7b00f34" />

**And it made no difference.** Here is the terminal log...

<img width="1082" height="622" alt="Screenshot 2025-11-11 165241" src="https://github.com/user-attachments/assets/c931a1ec-2b12-4647-9dcc-fe76ea0ffefb" />

The first run of `briefcase dev` is before I modified `dev.py`. 
The second time I executed it, line 35 was (definitely) commented out in 
`D:\beeware-tutorial\beeware-venv\Lib\site-packages\briefcase\commands\dev.py`

Let me know if I can assist you with any other tests.

**Reply by freakboy3742** [reply]
My apologies - I got the line number wrong. It's dev mode that we're looking to disable; by your line reference, that's line 33.

**Comment by aendie** [comment]
Don't apologise - I suspected it's the wrong line.
Yes, it works - no crash is reproducible with line 33 commented out :-)

**Reply by freakboy3742** [reply]
Excellent - glad we found a cause (or at least a workaround). Sounds like we might need to add some documentation to Toga and Briefcase to flag that Python's dev mode can be problematic for Winforms.

When the next Briefcase release lands, you'll be able to run `PYTHONDEVMODE=0 briefcase dev` to disable this behavior without manually tweaking the Briefcase source code.
